File: prettier-plugin-csharpier/src/Printer/Types/ConstructorDeclaration.ts
import { Doc } from "prettier";
import { printLeadingComments } from "../PrintComments";
import { printExtraNewLines } from "../PrintExtraNewLines";
import { PrintMethod } from "../PrintMethod";
import { printIdentifier, printModifiers, SyntaxToken, SyntaxTreeNode } from "../SyntaxTreeNode";
import { concat, group, hardline, indent, join, softline, line, doubleHardline } from "../Builders";
import { ArrowExpressionClauseNode } from "./ArrowExpressionClause";
import { AttributeListNode } from "./AttributeList";
import { BlockNode, printBlock } from "./Block";
import { ParameterListNode, printParameterList } from "./ParameterList";

class ConstructorInitializerNode {}

export interface ConstructorDeclarationNode extends SyntaxTreeNode<"ConstructorDeclaration"> {
    attributeLists: AttributeListNode[];
    modifiers: SyntaxToken[];
    identifier: SyntaxToken;
    parameterList?: ParameterListNode;
    initializer?: ConstructorInitializerNode;
    body?: BlockNode;
    expressionBody?: ArrowExpressionClauseNode;
    semicolonToken?: SyntaxToken;
}

export const printConstructorDeclaration: PrintMethod<ConstructorDeclarationNode> = (path, options, print) => {
    const node = path.getValue();
    const parts: Doc[] = [];
    printExtraNewLines(node, parts, "attributeLists", "modifiers", "identifier");
    printLeadingComments(node, parts, "attributeLists", "modifiers", "identifier");
    parts.push(printModifiers(node));
    parts.push(printIdentifier(node));
    parts.push(path.call(o => printParameterList(o, options, print), "parameterList"));
    parts.push(path.call(o => printBlock(o, options, print), "body"));

    return group(concat(parts));
};

File: Src/CSharpier/SyntaxPrinter/SyntaxNodePrinters/ConstructorDeclaration.cs
using System;
using CSharpier.DocTypes;
using CSharpier.SyntaxPrinter;
using Microsoft.CodeAnalysis.CSharp.Syntax;

namespace CSharpier.SyntaxPrinter.SyntaxNodePrinters
{
    public static class ConstructorDeclaration
    {
        public static Doc Print(ConstructorDeclarationSyntax node)
        {
            var groupId = Guid.NewGuid().ToString();

            return Doc.Group(
                ExtraNewLines.Print(node),
                new Printer().PrintAttributeLists(node, node.AttributeLists),
                Doc.Group(
                    Modifiers.Print(node.Modifiers),
                    Token.Print(node.Identifier),
                    ParameterList.Print(node.ParameterList, groupId),
                    Doc.IfBreak(Doc.Null, Doc.SoftLine)
                ),
                node.Initializer != null
                    ? Node.Print(node.Initializer)
                    : Doc.Null,
                node.Body != null
                    ? Block.PrintWithConditionalSpace(node.Body, groupId)
                    : Doc.Null,
                node.ExpressionBody != null
                    ? ArrowExpressionClause.Print(node.ExpressionBody)
                    : Doc.Null,
                Token.Print(node.SemicolonToken)
            );
        }
    }
}
